src/mpi/misc/utils.c

```diff
@@ -72,11 +63,21 @@ int MPIR_Localcopy(const void *sendbuf, MPI_Aint sendcount, MPI_Datatype sendtyp
         MPIR_ERR_CHKANDJUMP(actual_pack_bytes != copy_sz, mpi_errno, MPI_ERR_TYPE,
                             "**dtypemismatch");
     } else {
-        char *buf;
         intptr_t sfirst;
         intptr_t rfirst;
 
-        MPIR_CHKLMEM_MALLOC(buf, char *, COPY_BUFFER_SZ, mpi_errno, "buf", MPL_MEM_BUFFER);
+        MPL_gpu_query_pointer_type(sendbuf, &send_type);
+        MPL_gpu_query_pointer_type(recvbuf, &recv_type);
+
+        if (send_type == MPL_GPU_POINTER_DEV && recv_type == MPL_GPU_POINTER_DEV) {
+            MPL_gpu_dev_id_t devid;
+            MPL_gpu_get_device_id(&devid);
```

Comment:
What is this function?  Where is this getting a device ID from?

Reply:
Hmmm, right now it is considering the current GPU context. So in principle I see two cases: `1` - the `MPIR_Localcopy` is local, meaning we are moving data within the process address space (either D2D, D2H or H2D); `2` - the `MPIR_Localcopy` is not local, meaning we are moving data from another process address space (on a different GPU context) into the process address space (possibly to the GPU memory). In this case I guess allocating the temporary buffer in the target process GPU should be fine. The device id query function should probably be more generic still, also taking a buffer as input and returning the device id to which that buffer belongs to. Is this your concern here?

Reply:
What is "current GPU context"?  There is no such thing for L0 API.

Reply:
Besides, even for CUDA, "current GPU context" is irrelevant here.  The only thing that matters is which device my buffers map to.

Reply:
You are right. I will fix this.